src/features/machine/computer.js

```diff
@@ -182,29 +160,35 @@ function reportTiming(time) {
 
 const throttledReportTiming = throttle(reportTiming, 1000, {trailing: true })
 
-export const transformShapes = (state) => {
+export const transformShapes = (vertices, transform) => {
   const startTime = performance.now()
-  const input = getShapeVertices(state)
-  const numLoops = state.transform.repeatEnabled ? state.transform.numLoops : 1
-  const numTrackLoops = state.transform.repeatEnabled ? state.transform.trackNumLoops : 1
+  const numLoops = transform.repeatEnabled ? transform.numLoops : 1
+  const numTrackLoops = transform.repeatEnabled ? transform.trackNumLoops : 1
   let outputVertices = []
 
-  for (var i=0; i<numLoops; i++) {
-    if (state.transform.trackEnabled && numTrackLoops > 1) {
+  if (transform.trackEnabled && numTrackLoops > 1) {
+    for (var i=0; i<numLoops; i++) {
       for (var t=0; t<numTrackLoops; t++) {
-        outputVertices = outputVertices.concat(buildTrackLoop(state, i, t))
+        outputVertices = outputVertices.concat(buildTrackLoop(vertices, transform, i, t))
       }
-    } else {
-      for (i=0; i<numLoops; i++) {
-        for (var j=0; j<input.length; j++) {
-          let amount = state.transform.transformMethod === 'smear' ? i + j/input.length : i
-          outputVertices.push(transform(state.transform, input[j], amount, amount))
-        }
+    }
+  } else {
+    for (i=0; i<numLoops; i++) {
+      if (transform.transformMethod === 'smear' && transform.repeatEnabled) {
+        // remove last vertex; we don't want to return to our starting point when completing the shape
+        outputVertices.pop()
```

Comment:
This still isn't quite right.

Reply:
I think I fixed it. The startingSize wasn't being applied the right way to the tracks, and the non-smeared shapes weren't being connected the right way. There was also a (probably very old) but where I was iterating over i in a loop, then iterating over i again later, but it was somehow missed, because it was iterating over the same thing.